File: java/src/main/java/com/allstask/api/Metrics.java
// this file is @generated
package com.allstask.api;

import com.allstask.AllstaskHttpClient;
import com.allstask.Utils;
import com.allstask.exceptions.ApiException;
import com.allstask.models.CreateMetricRequest;
import com.allstask.models.Metric;
import com.allstask.models.MetricListResponse;
import com.allstask.models.UpdateMetricRequest;

import okhttp3.HttpUrl;

import java.io.IOException;

public class Metrics {
    private final AllstaskHttpClient client;

    public Metrics(AllstaskHttpClient client) {
        this.client = client;
    }

    /** */
    public MetricListResponse listMetrics() throws IOException, ApiException {

        return this.listMetrics(new MetricsListMetricsOptions());
    }

    /** */
    public MetricListResponse listMetrics(final MetricsListMetricsOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url = this.client.newUrlBuilder().encodedPath("/api/v1/metrics");
        if (options.productFamilyId != null) {
            url.addQueryParameter(
                    "product_family_id", Utils.serializeQueryParam(options.productFamilyId));
        }
        if (options.search != null) {
            url.addQueryParameter("search", options.search);
        }
        if (options.orderBy != null) {
            url.addQueryParameter("order_by", options.orderBy);
        }
        if (options.page != null) {
            url.addQueryParameter("page", Utils.serializeQueryParam(options.page));
        }
        if (options.perPage != null) {
            url.addQueryParameter("per_page", Utils.serializeQueryParam(options.perPage));
        }
        return this.client.executeRequest("GET", url.build(), null, null, MetricListResponse.class);
    }

    /** */
    public Metric createMetric(final CreateMetricRequest createMetricRequest)
            throws IOException, ApiException {
        HttpUrl.Builder url = this.client.newUrlBuilder().encodedPath("/api/v1/metrics");
        return this.client.executeRequest(
                "POST", url.build(), null, createMetricRequest, Metric.class);
    }

    /** */
    public Metric getMetric(final String metricId) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/metrics/%s", metricId));
        return this.client.executeRequest("GET", url.build(), null, null, Metric.class);
    }

    /** Partially update metric fields. Code and aggregation_type are immutable. */
    public Metric updateMetric(final String metricId, final UpdateMetricRequest updateMetricRequest)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/metrics/%s", metricId));
        return this.client.executeRequest(
                "PATCH", url.build(), null, updateMetricRequest, Metric.class);
    }

    /** */
    public void archiveMetric(final String metricId) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/metrics/%s/archive", metricId));
        this.client.executeRequest("POST", url.build(), null, null, null);
    }

    /** */
    public void unarchiveMetric(final String metricId) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/metrics/%s/unarchive", metricId));
        this.client.executeRequest("POST", url.build(), null, null, null);
    }
}

File: typescript/src/api/metrics.ts
// this file is @generated
import {
    CreateMetricRequest,
    CreateMetricRequestSerializer,
} from '../models/createMetricRequest';
import {
    Metric,
    MetricSerializer,
} from '../models/metric';
import {
    MetricListResponse,
    MetricListResponseSerializer,
} from '../models/metricListResponse';
import {
    ProductFamilyId,
    ProductFamilyIdSerializer,
} from '../models/productFamilyId';
import {
    UpdateMetricRequest,
    UpdateMetricRequestSerializer,
} from '../models/updateMetricRequest';
import { HttpMethod, AllstaskRequest, AllstaskRequestContext } from "../request";

export interface MetricsListMetricsOptions {
        productFamilyId?: ProductFamilyId;
        /** Search by metric name or code */
            search?: string;
        /** Sort order. Format: `column.direction`. Allowed columns: `name`, `code`, `created_at`. Direction: `asc` or `desc`. Default: `name.asc`. */
            orderBy?: string;
        /** Page number (0-indexed) */
            page?: number;
        /** Number of items per page */
            perPage?: number;
        }

    export class Metrics {
    public constructor(private readonly requestCtx: AllstaskRequestContext) {}

    /**  */
        public listMetrics(
            options?: MetricsListMetricsOptions,
            ): Promise<MetricListResponse> {
            const request = new AllstaskRequest(HttpMethod.GET, "/api/v1/metrics");

            request.setQueryParam("product_family_id", options?.productFamilyId);
            request.setQueryParam("search", options?.search);
            request.setQueryParam("order_by", options?.orderBy);
            request.setQueryParam("page", options?.page);
            request.setQueryParam("per_page", options?.perPage);
            
                return request.send(
                    this.requestCtx,
                    MetricListResponseSerializer._fromJsonObject,
                );
            }

        

    /**  */
        public createMetric(
            createMetricRequest: CreateMetricRequest,
            ): Promise<Metric> {
            const request = new AllstaskRequest(HttpMethod.POST, "/api/v1/metrics");

            request.setBody(
                    CreateMetricRequestSerializer._toJsonObject(
                        createMetricRequest,
                    )
                );
            
                return request.send(
                    this.requestCtx,
                    MetricSerializer._fromJsonObject,
                );
            }

        

    /**  */
        public getMetric(
            metricId: string,
            ): Promise<Metric> {
            const request = new AllstaskRequest(HttpMethod.GET, "/api/v1/metrics/{metric_id}");

            request.setPathParam("metric_id", metricId);
            
                return request.send(
                    this.requestCtx,
                    MetricSerializer._fromJsonObject,
                );
            }

        

    /** Partially update metric fields. Code and aggregation_type are immutable. */
        public updateMetric(
            metricId: string,
            updateMetricRequest: UpdateMetricRequest,
            ): Promise<Metric> {
            const request = new AllstaskRequest(HttpMethod.PATCH, "/api/v1/metrics/{metric_id}");

            request.setPathParam("metric_id", metricId);
            request.setBody(
                    UpdateMetricRequestSerializer._toJsonObject(
                        updateMetricRequest,
                    )
                );
            
                return request.send(
                    this.requestCtx,
                    MetricSerializer._fromJsonObject,
                );
            }

        

    /**  */
        public archiveMetric(
            metricId: string,
            ): Promise<void> {
            const request = new AllstaskRequest(HttpMethod.POST, "/api/v1/metrics/{metric_id}/archive");

            request.setPathParam("metric_id", metricId);
            
                return request.sendNoResponseBody(this.requestCtx);
            }

        

    /**  */
        public unarchiveMetric(
            metricId: string,
            ): Promise<void> {
            const request = new AllstaskRequest(HttpMethod.POST, "/api/v1/metrics/{metric_id}/unarchive");

            request.setPathParam("metric_id", metricId);
            
                return request.sendNoResponseBody(this.requestCtx);
            }

        

    }

